### PR TITLE
Ensure index updates do not fail if the local and remote history diverge

### LINKED ISSUE #199: Determine if index clone can be fast-forwarded, if not re-clone
/kind proposal
/priority important-soon

Once we start offering multiple indexes (#23), inevitably the plugin authors will do weird stuff like rewriting the git history (probably accidentally) and we don't want this to break `krew update`.

**Decision 1:** Should we try to handle this case?

So I think, ahead of time, we need to implement something that does:
1. git fetch-ish something to retrieve the commits (requires connection)
2. determine if the HEAD can be fast-forwarded (offline operation)
3. if can be fast-forwarded, go ahead (offline)
4. if cannot be fast-forwarded (at this point it's safe(?) to assume it's because someone rewrote the history), nuke the local clone and re-clone the repo.

**Decision 2 (if decision 1 is accepted):** Are the steps above a sound approach to this problem?

### PR DESCRIPTION
**Pull request by corneliusweig**
Unfortunately, this is not testable at the moment, but this should be covered in #198.

Close #199

### CODE REVIEW

File: pkg/gitutil/git.go [hunk 1]
```diff
@@ -47,7 +47,15 @@ func IsGitCloned(gitPath string) (bool, error) {
 
 // update will fetch origin and set HEAD to origin/HEAD.
 func update(destinationPath string) error {
-	return exec(destinationPath, "pull", "--ff-only", "-v")
+	if err := exec(destinationPath, "fetch", "-v"); err != nil {
+		return errors.Wrapf(err, "fetch index at %q failed", destinationPath)
+	}
+
+	if err := exec(destinationPath, "reset", "--hard", "@{upstream}"); err != nil {
+		return errors.Wrapf(err, "reset index at %q failed", destinationPath)
```
Comment (ahmetb) [inline pkg/gitutil/git.go]: you actually don't need to do err!=nil check for the last err, you can just return it (as errors.Wrapf will return nil when err==nil.

Comment (corneliusweig) [reply pkg/gitutil/git.go]: cool, I didn't know that :sunglasses:

**Pull request merged**